assets/tests/TaskSchemaValidationService.js

```diff
@@ -0,0 +1,225 @@
+  function(SYSTEM_CODE) {
+    var AUXILIARY_CODE_CLASS_NAME = 'AuxiliaryCode.';
+    var SYSTEM_CODE_CLASS_NAME = 'System.';
+    // TODO(eyurko): Update this once we support nonlinear runtimes.
+    var ALLOWED_RUNTIMES = ['linear'];
+    var _checkIfFunctionExistsInCode = function(functionName, code) {
+      var functionDefinition = 'def ' + functionName;
+      return code.indexOf(functionDefinition) !== -1;
+    };
+    var _checkIfFunctionExistsInClass = function(functionName, 
+      className, code) {
+      if (functionName.startsWith(className)) {
+        var strippedFunctionName = functionName.substring(
+          className.length);
+        return _checkIfFunctionExistsInCode(
+          strippedFunctionName, code);
+      }
+      return false;
+    };
+    var taskSchemaValidator = {
+      _starterCode: '',
+      _auxiliaryCode: '',
+      getAllFunctions: function() {
+        var functions = [];
+        for (var l in this) {
+          if (this.hasOwnProperty(l) && 
+              this[l] instanceof Function &&
+              !(/functions/i).test(l) &&
+              l.toString().startsWith('verify')) {
+            functions.push(this[l]);
+          }
+        }
+        return functions;
+      },
+      init: function(starterCode, auxiliaryCode) {
+        this._starterCode = starterCode;
+        this._auxiliaryCode = auxiliaryCode;
+      },
+      verifyInstructionsAreNotEmpty: function(task) {
+        var instructions = task.getInstructions();
+        return instructions.length > 0;
+      },
+      verifyInstructionsIsArrayOfStrings: function(task) {
+        var instructions = task.getInstructions();
+        return (angular.isArray(instructions) &&
+          instructions.every(function(instruction) {
+            return angular.isString(instruction);
+          }));
+      },
+      verifyMainFunctionNameIsString: function(task) {
+        var mainFunctionName = task.getMainFunctionName();
+        return (angular.isString(mainFunctionName) &&
+            mainFunctionName.length > 0);
+      },
+      verifyMainFunctionNameAppearsInStarterCode: function(
+        task) {
+        var mainFunctionName = task.getMainFunctionName();
+        return _checkIfFunctionExistsInCode(mainFunctionName, 
+          taskSchemaValidator._starterCode);
+      },
+      verifyInputFunctionNameIsNullOrString: function(task) {
+        var inputFunctionName = task.getInputFunctionName();
+        return (inputFunctionName === null ||
+          angular.isString(inputFunctionName));
+      },
+      verifyInputFunctionNameAppearsInAuxiliaryCodeIfNotNull: function(
+        task) {
+        var inputFunctionName = task.getInputFunctionName();
+        return (inputFunctionName === null ||
+          _checkIfFunctionExistsInClass(
+              inputFunctionName, AUXILIARY_CODE_CLASS_NAME, 
+              taskSchemaValidator._auxiliaryCode));
+      },
+      verifyOutputFunctionNameIsNullOrString: function(task) {
```

Comment:
Perhaps one thing to check for all these function names is that they start with a valid prefix. We have a small set of valid prefixes: ["AuxiliaryCode.", "System."]

I can easily see someone forgetting to add the class name to the function name.